hicommonwealth/commonwealth
# Global activity cache fixes

## Link to Issue
Closes: #5204 

## Description of Changes
- Recreates the global activity cache so it uses Redis.
  - This is important because an in-memory cache could lead to inconsistent refresh. For example (without Redis), on web dyno 1 a thread is deleted and the in-memory cache is refreshed but then the user refreshes the page and web dyno 2 responds. In this case, the cache on web dyno 2 is stale and therefore the thread still appears in the feed.
- Modified the `getActivityFeed` function so that it uses only a single query and only returns the data that is required to render the feed (small to medium performance gain depending on current DB load + distance to server)
  - Limits fetching profiles for each notification to the first 4 since only 4 avatars are displayed on the client
  - Joins with comments table to stop deleted comments from rendering in the feed
- Added a check in the thread and comment deletion logic which updates the cache if the deleted thread or comment is stored in it.
- Removed a bunch of unused attributes from the `DashboardActivityNotification` client model

## "How We Fixed It"

- Refresh the cache when a top 50 thread is deleted or when a comment on a top 50 thread is deleted

## Test Plan
**Test thread/comment deletion**
- Make yourself a super-admin
- Navigate to the global activity dashboard feed
- Click on a top thread
- Delete the thread
- Go back and refresh the feed (thread should no longer appear)
- Repeat the same but delete a comment

**Test cache + cache refresh**
- Start a fresh server (restart server)
- Create a thread
- View the feed
  - The thread should not show up in the feed
- Wait 5 minutes
- Refresh the feed and now the thread should show up in the feed

## Deployment Plan

1. 

## Other Considerations

-
